src/hyperlight_host/src/hypervisor/whp.rs

```diff
@@ -0,0 +1,410 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::os::raw::c_void;
+
+use hyperlight_common::mem::PAGE_SIZE_USIZE;
+use windows::core::s;
+use windows::Win32::Foundation::{FreeLibrary, HANDLE};
+use windows::Win32::System::Hypervisor::*;
+use windows::Win32::System::LibraryLoader::*;
+use windows_result::HRESULT;
+
+#[cfg(gdb)]
+use super::handlers::DbgMemAccessHandlerWrapper;
+use super::regs::{
+    AlignedRegisterValues, WHP_FPU_NAMES, WHP_FPU_NAMES_LEN, WHP_REGS_NAMES, WHP_REGS_NAMES_LEN,
+    WHP_SREGS_NAMES, WHP_SREGS_NAMES_LEN,
+};
+use super::vm::HyperlightExit;
+use super::wrappers::HandleWrapper;
+use crate::hypervisor::regs::{CommonFpu, CommonRegisters, CommonSpecialRegisters};
+use crate::hypervisor::surrogate_process::SurrogateProcess;
+use crate::hypervisor::surrogate_process_manager::get_surrogate_process_manager;
+use crate::hypervisor::vm::Vm;
+use crate::mem::memory_region::{MemoryRegion, MemoryRegionFlags};
+use crate::{new_error, Result};
+
+pub(crate) fn is_hypervisor_present() -> bool {
+    let mut capability: WHV_CAPABILITY = Default::default();
+    let written_size: Option<*mut u32> = None;
+
+    match unsafe {
+        WHvGetCapability(
+            WHvCapabilityCodeHypervisorPresent,
+            &mut capability as *mut _ as *mut c_void,
+            std::mem::size_of::<WHV_CAPABILITY>() as u32,
+            written_size,
+        )
+    } {
+        Ok(_) => unsafe { capability.HypervisorPresent.as_bool() },
+        Err(_) => {
+            log::info!("Windows Hypervisor Platform is not available on this system");
+            false
+        }
+    }
+}
+
+// This function dynamically loads the WHvMapGpaRange2 function from the winhvplatform.dll
+// WHvMapGpaRange2 only available on Windows 11 or Windows Server 2022 and later
+// we do things this way to allow a user trying to load hyperlight on an older version of windows to
+// get an error message saying that hyperlight requires a newer version of windows, rather than just failing
+// with an error about a missing entrypoint
+// This function should always succeed since before we get here we have already checked that the hypervisor is present and
+// that we are on a supported version of windows.
+type WHvMapGpaRange2Func = unsafe extern "cdecl" fn(
+    WHV_PARTITION_HANDLE,
+    HANDLE,
+    *const c_void,
+    u64,
+    u64,
+    WHV_MAP_GPA_RANGE_FLAGS,
+) -> HRESULT;
+
+/// A Hypervisor driver for HyperV-on-Windows.
+#[derive(Debug)]
+pub(crate) struct WhpVm {
+    partition: WHV_PARTITION_HANDLE,
+    // Lazily create the surrogate process when we need to map memory
+    surrogate_process: Option<SurrogateProcess>,
+    mmap_file_handle: HandleWrapper,
+}
+
+unsafe impl Send for WhpVm {}
+unsafe impl Sync for WhpVm {}
+
+#[repr(C, align(16))]
```

Comment:
Maybe add a comment here on why we do this :)

Reply:
fixed